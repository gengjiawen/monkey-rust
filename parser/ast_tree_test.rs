#[cfg(test)]
mod tests {
    use crate::parse;
    use insta::*;

    pub fn test_ast_tree(name: &str, input: &str) {
        let let_ast = match parse(input) {
            Ok(node) => {
                serde_json::to_string_pretty(&node).unwrap()
            }
            Err(e) => format!("parse error: {}", e[0])
        };
        println!("{}", let_ast);
        assert_snapshot!(name, let_ast);
    }

    #[test]
    fn test_let() {
        let input = "let a = 3";
        test_ast_tree("test_let", input)
    }

    #[test]
    fn test_string() {
        let input = r#""jw""#;
        test_ast_tree("test_string", input)
    }

    #[test]
    fn test_array() {
        let input = "[1, true]";
        test_ast_tree("test_array", input)
    }

    #[test]
    fn test_hash() {
        let input = r#"{"a": 1}"#;
        test_ast_tree("test_hash", input)
    }
}
