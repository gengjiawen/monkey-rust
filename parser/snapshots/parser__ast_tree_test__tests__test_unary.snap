---
source: parser/ast_tree_test.rs
expression: "-3"
---
{
  "Program": {
    "type": "Program",
    "body": [
      {
        "type": "UnaryExpression",
        "op": {
          "span": {
            "start": 0,
            "end": 1
          },
          "kind": {
            "type": "MINUS"
          }
        },
        "operand": {
          "type": "Integer",
          "raw": 3,
          "span": {
            "start": 1,
            "end": 2
          }
        },
        "span": {
          "start": 0,
          "end": 2
        }
      }
    ],
    "span": {
      "start": 0,
      "end": 3
    }
  }
}
